### PR TITLE
add documentation for mod migration and porting

### PR DESCRIPTION
**Pull request by TheElan**

### CODE REVIEW

File: docs/migration.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+   2. `src`
+   3. `LICENSE`
+   4. `README.md`
+   5. `build.gradle(.kts)`
+2. Copy all repositories from your `build.gradle(.kts)` to `repositories.gradle`
+3. Copy all dependencies from your `build.gradle(.kts)` to `dependecies.gradle`
+4. replace your `build.gradle(.kts)` with `build.gradle` from template. In case you have written some custom tasks/configurations not present in the template - you'll best to move them to separate file and link it in this, like for example it does this with `dependencies.gradle`
+5. Adapt `gradle.properties` to your mod
+6. Ensure `src/main/resources/mcmod.info` contains `${modId}`, `${modName}`. `${modVersion}` and `${minecraftVersion}`
+7. Run `./gradlew clean setupDecompWorkspace`
```
Comment (SinTh0r4s) [inline docs/migration.md:19]: Reimport the project before setupDecompWorkspace

Comment (SinTh0r4s) [reply docs/migration.md:19]: And what about .gitignore, GitHub CI and most importantly Jitpack CI?

File: docs/migration.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Migration guides
+
+## Generic migration
+Migration for the typical mod which doesn't usy anything special but minecraft forge and some library dependencies. 
+For core plugin, mixins, shadowing, access transformes, asm or etc. you'll need to do some extra steps. 
+If they are missing in this document - we will gladly receive your suggestions/contribution. 
+
+1. Copy and replace all files from template to your repository, but: // TODO: (maybe we should have a task in CI which attaches zip with all files but these, called "migration-bundle.zip" or smth)
```
Comment (SinTh0r4s) [inline docs/migration.md:8]: No idea how to do that, but it might be a nice idea

File: docs/migration.md (hunk 1 above)
Comment (TheElan) [reply docs/migration.md:19]: I've provided list of exclusions when coping, which should imply that all CI files and gitignore should be copied

File: docs/migration.md (hunk 2 above)
Comment (TheElan) [inline docs/migration.md:8]: maybe we should have a task in CI which attaches zip with all files but these, called "migration-bundle.zip", so it's just "copy and replace all files to your project's root"

File: README.md [hunk 3]
```diff
@@ -40,23 +43,24 @@ We had our fair share in struggles with build scripts for Minecraft Forge. There
...
 You may activate Forge's Access Transformers by defining a configuration file in `gradle.properties`.
 
 Check out the [`example-access-transformers`](https://github.com/SinTh0r4s/ExampleMod1.7.10/tree/example-access-transformers) brach for a working example!
 
-### For Advanced Modders: Mixins
+### Mixins
 
-You may activate Mixin in 'gradle.properties'. In that case a mixin configuration (usually named `mixins.mymodid.json`) will be generated automatically, and you only have to write the mixins itself. Dependencies are handled as well.
-Take a look at the examples in [`com.myname.mymodid.mixinplugin.*`](https://github.com/SinTh0r4s/ExampleMod1.7.10/tree/example-mixins/src/main/java/com/myname/mymodid/mixinplugin) and [`com.myname.mymodid.mixins.*`](https://github.com/SinTh0r4s/ExampleMod1.7.10/tree/example-mixins/src/main/java/com/myname/mymodid/mixins). The use of a MixinPlugin allows the injection into any mod and not just vanilla Minecraft.
+Mixins are usually used to modify vanilla or other mod/library from a mod level. For example, redirect a call, change visibility or make class implement your interface. It's an advanced topic and most mods don't need to do that.
```
Comment (SinTh0r4s) [inline README.md:54]: What are you trying to say with `mod level`?

Comment (TheElan) [reply README.md:54]: Like from within your mod, as in from your own sources, without touching sources of what you are modifing.

File: docs/migration.md (hunk 2 above)
Comment (TheElan) [reply docs/migration.md]: I think you can do this with actions we already have, just add a task which will zip all the things we want to and add it to the release, smth like:

```
    - name: Gather migration template files
      run: zip build/migrations.zip * --exclude build.gradle,src,...
        
    - uses: "marvinpinto/action-automatic-releases@latest"
      with:
        repo_token: "${{ secrets.GITHUB_TOKEN }}"
        automatic_release_tag: "latest"
        prerelease: true
        title: "Latest Development Build"
        files: [build/libs/*.jar, build/migrations.zip]
```
This is a pseudo code, I don't remember how to exclude things in zip command and am not exactly sure how to specify files in the actions.

File: README.md [hunk 4]
```diff
@@ -40,23 +43,24 @@ We had our fair share in struggles with build scripts for Minecraft Forge. There
...
 
-### For Advanced Modders: Mixins
+### Mixins
 
-You may activate Mixin in 'gradle.properties'. In that case a mixin configuration (usually named `mixins.mymodid.json`) will be generated automatically, and you only have to write the mixins itself. Dependencies are handled as well.
-Take a look at the examples in [`com.myname.mymodid.mixinplugin.*`](https://github.com/SinTh0r4s/ExampleMod1.7.10/tree/example-mixins/src/main/java/com/myname/mymodid/mixinplugin) and [`com.myname.mymodid.mixins.*`](https://github.com/SinTh0r4s/ExampleMod1.7.10/tree/example-mixins/src/main/java/com/myname/mymodid/mixins). The use of a MixinPlugin allows the injection into any mod and not just vanilla Minecraft.
+Mixins are usually used to modify vanilla or mod/library in runtime without having to change source code. For example, redirect a call, change visibility or make class implement your interface. It's an advanced topic and most mods don't need to do that.
+
+To activate Mixin in 'gradle.properties'. In that case a mixin configuration (usually named `mixins.mymodid.json`) will be generated automatically, and you only have to write the mixins itself. Dependencies are handled as well.
```
Comment (SinTh0r4s) [inline README.md:56]: Not quite a sentence: To activate Mixin in 'gradle.properties'.

File: docs/porting.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+Try to get rid of dependencies on concrete jars (usually in the `lib` folder) if any present. This way it will be much easier to change (upgrade/downgrade) your project dependencies, when needed.
+
+Check if they are in maven repository (usually authors put such information in the project readme), if it isn't accessible but project is open source with permissive enough license (e.g. MIT) - you still can publish them yourselves:
+   1. Fork the repository
+   2. Drop `jitpack.yml` and `.github/workflows/gradle.yml` in project root. You can find this file in this repository root.
+   3. Make sure everything builds from console by running `./gradlew clean setupCIWorkspace`
+   4. If all is fine/after fixing the errors - make a tag on Github or using console, this should trigger Github build hook and generate a release
+   4. Lookup forked repository on `https://jitpack.io/`
+   5. Find your release and click "Get it", which should scroll you down to the example of how to add the dependency (make sure you have jitpack repository in mod you are porting)
+   6. Checkout build log beside button you clicked to make sure it succeed
```
Comment (SinTh0r4s) [inline docs/porting.md:20]: make sure it __succeeds__

File: README.md (hunk 4 above)
Comment (TheElan) [reply README.md:56]: wired, probably somehow chopped some text when editing

File: docs/porting.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+Now, having dependency string you can just plug it into dependencies
```
Comment (SinTh0r4s) [inline docs/porting.md:22]: What are you trying to say?

File: docs/porting.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+If you can't find any source code of the lib, don't despair, there is still a chance to distribute it. If original author has not restricted the distribution permissions (hence the jar ended up in repository of the mod you are porting) or you managed to obtain their permission - you can just upload it as a jar to jitpack, see [jitpack single file publishing thread](https://gist.github.com/jitpack-io/f928a858aa5da08ad9d9662f982da983) for that.
```
Comment (SinTh0r4s) [inline docs/porting.md:24]: If __the__ original author

~~for that.~~

Are you trying to write about upgrading files in /libs to maven dependencies? It's not clear what you are talking about

File: docs/porting.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+There also may be a case where mod depends on another mod - then you'll need to port it first. (Yay, dependency hell! :D)
```
Comment (SinTh0r4s) [inline docs/porting.md:26]: There may also be a case where mods depend on other mods - then you'll need to port all dependencies first.

File: docs/porting.md (hunk 6 above)
Comment (TheElan) [reply docs/porting.md:22]: > Now when you are sure dependency is available in maven repository - just add it as a normal gradle dependency in `dependencies.gradle`

Is that more clear?

File: docs/porting.md (hunk 7 above)
Comment (TheElan) [reply docs/porting.md:24]: Yes, exactly what I am suggesting to do

File: docs/migration.md (hunk 2 above)
Comment (SinTh0r4s) [reply docs/migration.md]: We need to get this task in a separate file thought. Since the current file is supposed to be part of the pack.

**Pull request merged**